### PR TITLE
Externalize resource quota admission controller & controller reconciliation

### PR DESCRIPTION
**Pull request by yue9944882**
ref: #66680

This pull externalizes resource quota admission controller and also propagates externalization to quota controller:

1. Externalizes clientset/informers for resourcequota ad controller
2. _[API-CHANGE]_ For private schema of quota, we switch `MatchScopes` to external
3. Update generated stuff #67871
4. Copy & refactor an external pkg `pkg/quota` -> `pkg/quota/v1`
    a. Refactor `Evaluator` interface to accept/return external types
    b. Switching to externals in `pvc`/`pod`/`svc` evaluators
    c. Prune `pkg/quota`
5. For quota controller (`pkg/controller/quota`), we prune odd flipping object conversions between internal and external 🎉
6. Fixes pkg imports across project

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by yue9944882** [comment]
/hold

will sort them into logical commits once it's ready for a review
have to be super careful about this pull ⚠️

**Comment by yue9944882** [comment]
@kubernetes/sig-testing 🤕folks, rs test is flaking again. it's getting frequent recently.

**Comment by BenTheElder** [comment]
Please file a bug against that test and the owning sig. Looks like sig-apps given the shell path?  Not familiar with the rs tests.

File: plugin/pkg/admission/resourcequota/apis/resourcequota/types.go [hunk 1]
```diff
@@ -68,5 +68,5 @@ type LimitedResource struct {
 	// "PriorityClassNameIn=cluster-services"
 	// +optional
 	//	MatchScopes []string `json:"matchScopes,omitempty"`
-	MatchScopes []core.ScopedResourceSelectorRequirement `json:"matchScopes,omitempty"`
+	MatchScopes []corev1.ScopedResourceSelectorRequirement `json:"matchScopes,omitempty"`
```
Comment (deads2k) [inline plugin/pkg/admission/resourcequota/apis/resourcequota/types.go]: I didn't expect this.  One the other hand, I'm hard pressed to say why it shouldn't be allowed.

@liggitt we have anything that speaks against this?

File: pkg/quota/v1/resources_test.go [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+package quota
+
+import (
+	"testing"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+func TestEquals(t *testing.T) {
```
Comment (deads2k) [inline pkg/quota/v1/resources_test.go]: I'm surprised this is a copy, not a move.  Is there an original somewhere?  Looks like the file it's testing has completely moved.

File: pkg/quota/v1/evaluator/core/services.go [hunk 3]
```diff
@@ -144,12 +144,12 @@ func (p *serviceEvaluator) UsageStats(options quota.UsageStatsOptions) (quota.Us
 var _ quota.Evaluator = &serviceEvaluator{}
 
 //GetQuotaServiceType returns ServiceType if the service type is eligible to track against a quota, nor return ""
-func GetQuotaServiceType(service *v1.Service) v1.ServiceType {
+func GetQuotaServiceType(service *corev1.Service) corev1.ServiceType {
```
Comment (deads2k) [inline pkg/quota/v1/evaluator/core/services.go]: This I did not expect.  @derekwaynecarr Does this break evaluation when received via admission?

Actually, are we sure this works for both the admission side and the reconciliation side today?

File: pkg/quota/v1/evaluator/core/services.go [hunk 4]
```diff
@@ -73,65 +73,65 @@ func (p *serviceEvaluator) Handles(a admission.Attributes) bool {
...
-func (p *serviceEvaluator) UncoveredQuotaScopes(limitedScopes []api.ScopedResourceSelectorRequirement, matchedQuotaScopes []api.ScopedResourceSelectorRequirement) ([]api.ScopedResourceSelectorRequirement, error) {
-	return []api.ScopedResourceSelectorRequirement{}, nil
+func (p *serviceEvaluator) UncoveredQuotaScopes(limitedScopes []corev1.ScopedResourceSelectorRequirement, matchedQuotaScopes []corev1.ScopedResourceSelectorRequirement) ([]corev1.ScopedResourceSelectorRequirement, error) {
+	return []corev1.ScopedResourceSelectorRequirement{}, nil
 }
 
 // convert the input object to an internal service object or error.
-func toInternalServiceOrError(obj runtime.Object) (*api.Service, error) {
-	svc := &api.Service{}
+func toExternalServiceOrError(obj runtime.Object) (*corev1.Service, error) {
```
Comment (deads2k) [inline pkg/quota/v1/evaluator/core/services.go]: Ah, it's this function that does the magic.

**Comment by deads2k** [comment]
Looks equivalent. Just the API question.  Thinking about it, I'm inclined to say the API choice makes sense.

**Comment by deads2k** [comment]
Next time please squash your import changes where you moved the files.  Given the overall size of this pull, I'll allow it here, but if you have to rebase, please squash it in.

/lgtm

**Comment by liggitt** [comment]
helper change under apis lgtm

/approve

**Comment by yue9944882** [comment]
> Next time please squash your import changes where you moved the files. 

copy that. was trying to make it cleaner for a review.

File: pkg/quota/resources_test.go [hunk 5]
```diff
@@ -1,321 +0,0 @@
...
-
-package quota
-
-import (
-	"testing"
-
-	"k8s.io/apimachinery/pkg/api/resource"
-	api "k8s.io/kubernetes/pkg/apis/core"
-)
-
```
Comment (yue9944882) [inline pkg/quota/resources_test.go]: > Is there an original somewhere? Looks like the file it's testing has completely moved.

it's here originally 🙂

File: pkg/quota/v1/resources_test.go (hunk 2 above)
Comment (yue9944882) [reply pkg/quota/v1/resources_test.go]: yeah. am surprised too. 😅originally i didn't try to remove the old quota pkg. but after i finished switching controller part, the entire `pkg/quota` can vanish with a `rm -rf`.

File: plugin/pkg/admission/resourcequota/apis/resourcequota/types.go (hunk 1 above)
Comment (yue9944882) [reply plugin/pkg/admission/resourcequota/apis/resourcequota/types.go]: @deads2k @liggitt if you notice, in `v1alpha1` and `v1beta1` this field is external but this one isn't. looks really odd to me

File: pkg/quota/v1/evaluator/core/services.go (hunk 3 above)
Comment (derekwaynecarr) [reply pkg/quota/v1/evaluator/core/services.go]: isnt this pr changing both admission and reconciliation side to always use external?

Comment (deads2k) [reply pkg/quota/v1/evaluator/core/services.go]: > isnt this pr changing both admission and reconciliation side to always use external?

@derekwaynecarr yeah. See the switch of the helper function to external as well.

File: pkg/quota/v1/evaluator/core/services.go [hunk 6]
```diff
@@ -19,31 +19,31 @@ package core
...
-	api.ResourceServices,
-	api.ResourceServicesNodePorts,
-	api.ResourceServicesLoadBalancers,
+	corev1.ResourceServices,
+	corev1.ResourceServicesNodePorts,
+	corev1.ResourceServicesLoadBalancers,
 }
 
 // NewServiceEvaluator returns an evaluator that can evaluate services.
 func NewServiceEvaluator(f quota.ListerForResourceFunc) quota.Evaluator {
```
Comment (mengqiy) [inline pkg/quota/v1/evaluator/core/services.go]: > quota.ListerForResourceFunc

It has been renamed from "k8s.io/kubernetes/pkg/quota" to "k8s.io/kubernetes/pkg/quota/v1".
I'm really surprised to see this package name, `quota`, still work here w/o declaring a package name alias!
@yue9944882 any idea how this work?

Comment (yue9944882) [reply pkg/quota/v1/evaluator/core/services.go]: >  "k8s.io/kubernetes/pkg/quota/v1".

b/c we declared `package quota` in the source under `"k8s.io/kubernetes/pkg/quota/v1"`, we can call this an oversight when copy/pasting the quota codes.

Comment (yue9944882) [reply pkg/quota/v1/evaluator/core/services.go]: is that bothering you?

**Pull request merged**